src/plots/plots.js

```diff
@@ -479,6 +479,24 @@ plots.supplyDefaults = function(gd, opts) {
     // relink functions and _ attributes to promote consistency between plots
     relinkPrivateKeys(newFullLayout, oldFullLayout);
 
+    // For persisting GUI-driven changes in layout
+    // _preGUI and _tracePreGUI were already copied over in relinkPrivateKeys
+    if(!newFullLayout._preGUI) newFullLayout._preGUI = {};
+    // track trace GUI changes by uid rather than by trace index
```

Comment:
👀 @chriddyp 

This is important, as it has implications for how you use this feature: GUI changes to a trace (mostly just `visible: 'legendonly'` and `selectedpoints` but there's also `dimensions[i].constraintrange` in `parcoords`, and `name` and `colorbar.title` in editable mode) are tracked by the trace's `uid`, NOT by index.

That doesn't mean it will break if you don't provide a `uid` - since #2681 `uid` is provided consistently from call to call, so without `uid` we'll implicitly fall back on index. But if you want to ensure visibility follows traces of the same *meaning*, even if they change index, just provide a `uid` for that trace.

This could happen for example if you have a multiselect box, or bank of checkboxes, for which traces to include in a plot, and the user selects or deselects an earlier item after already having selected a later one.

Reply:
Could you add a few lines about this `uid` vs trace index logic in the `plots/attributes.js` uirevision description?